### PR TITLE
provide hint about what isn't hot reloadable

### PR DESCRIPTION
**Pull request by sethladd**

### CODE REVIEW

**Comment by sethladd** [comment]
Help with https://github.com/flutter/flutter/issues/12095

**Comment by sethladd** [comment]
cc @mit-mit

File: packages/flutter_tools/templates/create/lib/main.dart.tmpl [hunk 1]
```diff
@@ -13,6 +13,7 @@ void main() {
   // See https://flutter.io/testing/ for more info.
   enableFlutterDriverExtension();
 {{/withDriverTest}}
+  // Code directly inside of main() is NOT hot reloadable.
```
Comment (sethladd) [inline packages/flutter_tools/templates/create/lib/main.dart.tmpl]: maybe we should provide a link to explain why?

I don't think we have such a link, but could add one later.

**Comment by Hixie** [comment]
From #12095:

> Yeah, that's why the template app has such a small void main, to push people away from expecting it to hot reload.
>
> We could add a comment there but I'm concerned about the negative effect of having a negative comment ("hot reload doesn't work here") so early in the file, before any positive comments ("hot reload works here!").

I'm reluctant to add this (especially as written with an all-caps "NOT" and so forth) so early in the file.

A better solution might be to have a page that explains how hot reload actually works, and link to that from some of the early tutorial materials, rather than putting it here?

I think we can really impact people's enjoyment if we put limitations in their face so early.

Another option would be to phrase it differently. Something like "To rerun the main() function, which only runs once on startup, use a hot restart (rather than a hot reload)." for example. Though that as written here is still probably a little too negative?

**Comment by sethladd** [comment]
@Hixie thanks for the review. I've updated the comment based on your suggestion.

**Comment by mit-mit** [comment]
I'm still not convinced this belongs in our template. Will add more details Monday.

**Comment by sethladd** [comment]
Ideally, we'd make any code hot reloadable, so we don't need comments like this :/

In the meantime, we add comments to our template explaining how to use our framework, and it's been really helpful. It's 100% non-obvious why you don't see hot reloadable changes in main() and I've watched people struggle here. A hint/note seems appropriate, similar to how we add a hint about what setState() is. Not clear what the different is.

**Comment by Hixie** [comment]
Yeah as I said above I think the sentence you have now is still too negative to really be worth adding.

I don't really see how we could make `main()` hot reloadable. Hot reload doesn't re-run any code, it just changes the code on the fly. This code has already run, it only executes once per application run. (That's why a restart runs it again.)

I'm curious to hear @mit-mit's more detailed views.

**Comment by sethladd** [comment]
@Hixie the new version is still negative? I liked your version, it's pretty clear.

**Comment by mit-mit** [comment]
I have both general feedback (about instructional comments) and specific ones (about hot reload)

Generally, I agree that the initial comments we added were helpful and have aided with understanding things such as `setState`. But comments aren't "free", they are additional information that the developer has to read, understand, and absorb, so we should think carefully about any we add. I feel we have may gone a bit overboard already; our template file is now 111 lines, and out of those almost half (47) are now quite dense, information-loaded comments. cc @lukechurch do we have some UX research that might help is level-set here?

Specifically about hot reload, I think the core challenge is that the basics of hot reload are easily understood with basic intuition, but the details of what is possible **and desirable** to reload is actually not that easy to understand. The above review discussion about whether hot reload couldn't just be "fixed" to hot reload main changes is a good example of this. Understanding the different between reloading code vs reloading state is another.

At least in the specific case, and in fact in the general case, what I would really like to see us do, is invest in an actual onboarding experience on Flutter.io. This would walk the new developer through a set of small steps, each step educating them about a single core concept of what it means to develop a Flutter app. Each step would explain a concept, and show concrete changes the developer would do to the app template to achieve a single effect. The steps could be something like:

1. Installing Flutter
1. Creating a new app
1. Everything is a Widget
1. Making a change and hot reloading it
1. Layout out widgets with containers (Rows and Columns)
1. Managing steps, and updating state with setState()

And potentially continue onto a more advanced part two:

1. Debugging visual layout (debut paint, future => Inspector)
1. Creating platform-specific UI
1. Adding a package with ready-made functionality
1. Creating an animation
1. Making brand-specific layout
1. Understanding hot reload vs hot restarts

**Pull request closed without merging**